pkg/yurthub/proxy/util/util.go

```diff
@@ -295,3 +300,31 @@ func WithRequestTimeout(handler http.Handler) http.Handler {
 		handler.ServeHTTP(w, req)
 	})
 }
+
+func WithSaTokenSubstitute(handler http.Handler, tenantMgr tenant.Interface) http.Handler {
```

Comment:
i think that the logic of this handler should be as following:
```
if request has bearer token {
     // parse bearer token to resolve namespace

     // verify namespace in bearer token is tenant namespace or not.
     // if equal to tenant namespace, do not replace bearer token, goto handler.ServeHTTP(w, req)

     // wait for tenant token is synced

    // after tenant token synced, replace bearer token with tenant token in request header.
}
```

Reply:
I think there's no need to invoke  WaitForCacheSync func in handler, the reason is as follows:
1. waitforcachesync is just used to identify that all items in fifo have been pop to callback funcs,  it check no need to do it in every request handling.
2. yurthub will get the tenant token at its first start, if the net connection is down, informer will  start its list-watch from  
cacheManager's cache, it can still get the tennat token, and I have test disconnection scenario

If WaitForCacheSync must been invoked, the time should be at yurthub initialization, the informerFactory do the waitforcachesync for one time.



waitforcachesync will call 
`// HasSynced returns true if an Add/Update/Delete/AddIfNotPresent are called first,
// or the first batch of items inserted by Replace() has been popped.
func (f *DeltaFIFO) HasSynced() bool {
	f.lock.Lock()
	defer f.lock.Unlock()
	return f.populated && f.initialPopulationCount == 0
}`

Reply:
@MrGirl if `WaitForCacheSync` is not used, we can not sure that tenant token is ready. requests of flanneld/kube-proxy/coredns maybe come into handler before edge-hub's requests.